### PR TITLE
EZP-25578: Undefined offset when trying to update ContentType

### PR DESCRIPTION
**Pull request by clash82**
Link to Jira task: https://jira.ez.no/browse/EZP-25578

### CODE REVIEW

**Comment by clash82** [comment]
ping @bdunogier @andrerom @SylvainGuittard

**Comment by SylvainGuittard** [comment]
I confirm that this ```undefined offset``` issue occurs when we are updating a content type using the API.

File: eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php [hunk 1]
```diff
@@ -214,6 +214,11 @@ public function extractContentFromRows(array $rows, array $nameRows)
         $results = array();
         foreach ($contentInfos as $contentId => $contentInfo) {
             foreach ($versionInfos[$contentId] as $versionId => $versionInfo) {
+                // ignore missing versions
+                if (!array_key_exists($versionInfo->versionNo, $versionedNameData[$contentId])) {
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php]: `isset`

File: eZ/Publish/Core/Persistence/Legacy/Content/Type/ContentUpdater/Action/AddField.php [hunk 2]
```diff
@@ -96,6 +96,11 @@ function ($versionNo) use ($contentId) {
         foreach ($versionNumbers as $versionNo) {
             $contentRows = $this->contentGateway->load($contentId, $versionNo);
             $contentList = $this->contentMapper->extractContentFromRows($contentRows, $nameRows);
+
+            if (!array_key_exists(0, $contentList)) {
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/Content/Type/ContentUpdater/Action/AddField.php]: `isset`

**Comment by andrerom** [comment]
Could you tell a bit more about the missing version name data? Any clues?

**Comment by clash82** [comment]
> Could you tell a bit more about the missing version name data? Any clues?

It seems the new stack is more strict than legacy system. Legacy was handling those cases without error. We cannot say for sure why we have this situation. When we started ez.no migration we've discovered lots of junk data in database, probably because of the older migrations.

**Comment by andrerom** [comment]
Ok, could you check how it works by having a fallback instead of omitting the version?

Basically the whole patch will then be:
```diff
diff --git a/eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php b/eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php
index c0100b7..d59d65b 100644
--- a/eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php
+++ b/eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php
@@ -212,9 +212,16 @@ class Mapper
         $results = array();
         foreach ($contentInfos as $contentId => $contentInfo) {
             foreach ($versionInfos[$contentId] as $versionId => $versionInfo) {
+                // Fallback to just main language name if versioned name data is missing
+                if (isset($versionedNameData[$contentId][$versionInfo->versionNo])) {
+                    $names = $versionedNameData[$contentId][$versionInfo->versionNo];
+                } else {
+                    $names = [$contentInfo->mainLanguageCode => $contentInfo->name];
+                }
+
                 $content = new Content();
                 $content->versionInfo = $versionInfo;
-                $content->versionInfo->names = $versionedNameData[$contentId][$versionInfo->versionNo];
+                $content->versionInfo->names = $names;
                 $content->versionInfo->contentInfo = $contentInfo;
                 $content->fields = array_values($fields[$contentId][$versionId]);
                 $results[] = $content;
```

**Comment by clash82** [comment]
I can confirm that it's working. Should I apply this patch instead of my commit?

**Comment by andrerom** [comment]
> I can confirm that it's working. Should I apply this patch instead of my commit?

yes please, it is less destructive approach ;)  *(but we should also at some point investigate having a script to generate missing names to fix the data inconsistency issue)*

**Comment by clash82** [comment]
Updated. Thank you @andrerom for help :)

**Comment by bdunogier** [comment]
+1

**Comment by andrerom** [comment]
+1

**Comment by andrerom** [comment]
@bdunogier I guess we should merge this to 6.1+?

**Comment by andrerom** [comment]
> @bdunogier I guess we should merge this to 6.1+?

assuming yes

Fixed in 2c7012546ddd09375a37d97be844273ed3113a6

**Comment by bdunogier** [comment]
6.1+ is perfect yes.

**Pull request closed without merging**